rust/ql/lib/codeql/rust/internal/PathResolution.qll

```diff
@@ -1563,6 +1667,41 @@ module TraitIsVisible<relevantTraitVisibleSig/2 relevantTraitVisible> {
   }
 }
 
+pragma[nomagic]
+private predicate isMacroExpansion(AstNode expansion, Path macroDefPath) {
+  exists(MacroCall mc |
+    expansion = mc.getMacroCallExpansion() and
+    macroDefPath = mc.getPath()
+  )
+  or
+  exists(ItemNode adt |
+    expansion = adt.(Adt).getDeriveMacroExpansion(_) and
+    macroDefPath = adt.getAttr("derive").getMeta().getPath()
+  )
+}
+
+pragma[nomagic]
+predicate isInMacroExpansion(Path macroDefPath, AstNode n) {
+  isMacroExpansion(n, macroDefPath)
+  or
+  isInMacroExpansion(macroDefPath, n.getParentNode())
+}
+
+/**
+ * Holds if `n` is inside a macro expansion, and the macro _may_ originate from
+ * crate `crate`.
+ *
+ * The reason why we cannot be sure is that we need to consider all ancestor macro
+ * calls.
+ */
+pragma[nomagic]
+predicate isInMacroFromCrateExpansion(CrateItemNode crate, AstNode n) {
+  exists(Path macroDefPath |
+    isInMacroExpansion(macroDefPath, n) and
+    crate.getASourceFile().getFile() = resolvePathCand(macroDefPath).getFile()
```

Comment:
This predicate is only used with `n` being a `$crate` path. Inlining that as below changes the predicate size from 1.354.637 to 4.683 on the path resolution test DB. 

Making it a predicate with result might also be more natural?

```suggestion
CrateItemNode getDollarCratePathCrate(RelevantPath dollarCratePath) {
  dollarCratePath.isDollarCrate() and
  exists(Path macroDefPath |
    isInMacroExpansion(macroDefPath, dollarCratePath) and
    result.getASourceFile().getFile() = resolvePathCand(macroDefPath).getFile()
```